File: findFirstLastElement.py
'''
Time Complexity: O(log n) as O(2 log n ) -> becoz of two binary searches = O(log n)
Space Complexity: O(1) Constant Space as no extra space is being Allocated
Leetcode: Two-Function: Sucess! One-function: not Successful!

//Problems: Facing problem with One-function optimization: my idea is to store it in a list and then return the list but it is currently
storing only one value, as shown in the output.

|-> I'd really appreciate if you could point me to direction or point out where I might be going wrong?

'''


class Solution:
    def searchRange(self, nums, target):

        # Sanity check to check whether the list passed is not None or empty
        if nums == None or len(nums)==0:
            return [-1,-1]

        # Check-point for whether the target lies in the range of the list
        if nums[0] > target or nums[len(nums)-1]<target:
            return [-1,-1]

        first = self.BSfirst(nums, target)
        last = self.BSlast(nums, target)
        return [first,last]

    def BSfirst(self,nums,target):
        low = 0
        high = len(nums) - 1
        while low<=high:
            mid = low + (high - low)//2
            if(nums[mid]==target):
                if (mid==0 or nums[mid]>nums[mid-1]):
                    return mid
                else:
                    high = mid - 1
            elif nums[mid]>target:
                high = mid - 1
            else:
                low = mid + 1
        return -1

    def BSlast(self, nums, target):
        low = 0
        high = len(nums) - 1
        while low <= high:
            mid = low + (high - low) // 2
            if (nums[mid] == target):
                if (mid == len(nums)-1 or nums[mid] < nums[mid + 1]):
                    return mid
                else:
                    low = mid + 1
            elif nums[mid]<target:
                low = mid + 1
            else:
                high = mid - 1
        return -1
"""
ONE FUNCTION OPTIMIZATION:
 
class Solution:
    def searchRange(self, nums, target):

        # Sanity check to check whether the list passed is not None or empty
        if nums == None or len(nums)==0:
            return [-1,-1]

        # Check-point for whether the target lies in the range of the list
        if nums[0] > target or nums[len(nums)-1]<target:
            return [-1,-1]

        low = 0
        high = len(nums) - 1
        r = []
        while low<=high:
            mid = low + (high - low)//2
            if(nums[mid]==target):
                if mid==0 or nums[mid]>nums[mid-1]:
                    r.append(mid)
                else:
                    high = mid - 1

                if mid==len(nums)-1 or nums[mid]<nums[mid+1]:
                    r.append(mid)
                else:
                    low = mid - 1
            elif nums[mid]>target:
                high = mid - 1
            else:
                low = mid + 1
            return r
        return -1
        
OUTPUT: The target  7  is found at  [2]
"""

nums = [5,7,7,8,8,10]
target = 7
'''
nums = [5,7,7,8,8,10]
target = 6
nums = []
target = 0
'''
s = Solution()
pos = s.searchRange(nums,target)
print("The target ",target," is found at ",pos)

File: FindFirstLastElement.java
class Solution {
    
    private int firstbinary(int[] nums,int low, int high,int target){
        while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == 0 || nums[mid] > nums[mid - 1]){
                    return mid;
                }
                else{
                    high = mid -1;
                }
            }
            else if(nums[mid] > target){
                 high = mid - 1;
             }
             else{
                 low = mid + 1;
             }
            
            
            
        }
        return -1;
    }
    
    private int secondbinary(int[] nums, int low, int high, int target){
         while(low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == nums.length-1 || nums[mid] < nums[mid + 1]){
                    return mid;
                }
                else{
                    low = mid + 1;
                }
               
            }
             else if(nums[mid] > target){
                 high = mid - 1;
             }
             else{
                 low = mid + 1;
             }
            
            
        }
        return -1;
    }
    public int[] searchRange(int[] nums, int target) {
     
        int low = 0;
        int high = nums.length -1;
        if(nums == null || nums.length == 0 ) return new int[]{-1,-1};
        if(nums[0] > target || nums[high] < target) return new int[] {-1,-1};
        int first = firstbinary(nums,low,high,target);
        if(first == -1) return new int [] {-1,-1};
        int second = secondbinary(nums,first,high,target);
         if(second == -1) return new int [] {-1,-1};
        return new int [] {first,second};
    }
}


// Time Complexity : O(logn) + O(logn) = O(2logn) -> 2 is constant => O(logn)
// Space Complexity : O(1). Not using any extra Space
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : java syntax, edge cases
// Your code here along with comments explaining your approach

//Here need to find the first and last positions of the element. we can use binary search approach but we cant find first and last element in a single Binary search. So finding first number in first Binary Search and second number in second binary search. Then the TC will be O(logn) + O(logn) which is O(2logn) Here 2 is constant. So, O(logn). 
// Handling Edge Cases : If our mid is in the first Index, we cant compare mid with mid - 1 bcz it will throw outof the bound error. Same with nums.length-1.
// Optimization : After finding first element Index we can use it as low for finding second index. If we dont get first number then no need to find second number, return new int[]{-1,-1} there itself.
